### PR TITLE
[node] make crypto.randomFill[Sync] return type equal to first arg type

### PR DESCRIPTION
**Pull request by kjin**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/nodejs/node/pull/21550

Also see: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/26655#issuecomment-400384604

### CODE REVIEW

File: types/node/index.d.ts [hunk 1]
```diff
@@ -5894,10 +5894,10 @@ declare module "crypto" {
     export function pseudoRandomBytes(size: number): Buffer;
     export function pseudoRandomBytes(size: number, callback: (err: Error | null, buf: Buffer) => void): void;
 
-    export function randomFillSync(buffer: ArrayBufferView, offset?: number, size?: number): ArrayBufferView;
-    export function randomFill(buffer: ArrayBufferView, callback: (err: Error | null, buf: ArrayBufferView) => void): void;
-    export function randomFill(buffer: ArrayBufferView, offset: number, callback: (err: Error | null, buf: ArrayBufferView) => void): void;
-    export function randomFill(buffer: ArrayBufferView, offset: number, size: number, callback: (err: Error | null, buf: ArrayBufferView) => void): void;
+    export function randomFillSync<T extends Buffer|ArrayBufferView>(buffer: T, offset?: number, size?: number): T;
```
Comment (ikokostya) [inline types/node/index.d.ts]: style note: can you add spaces around `|` for consistency with other code.

**Comment by ZaneHannanAU** [comment]
Should just add realistically. I get that this was *technically* caused by my change but just add a signature rather than changing the signature entirely.

Also because of how ArrayBufferView is defined it's fine with Buffer so there's no need for it specifically.

```
export function randomFill<T extends ArrayBufferView>(buf: T, offset?: number, size?: number, callback: (err: Error | null, buf: T) => void): void;
export function randomFill(buf: Buffer, offset?: number, size?: number, callback: (err: Error | null, buf: Buffer) => void): void;
export function randomFillSync<T extends ArrayBufferView>(buf: T, offset?: number, size?: number): T;
export function randomFillSync(buf: Buffer, offset?: number, size?: number): Buffer;
```
Realistically you could revert then add those where appropriate.

**Review by Flarna — APPROVED** [review]

File: types/node/index.d.ts (hunk 1 above)
Comment (Flarna) [reply types/node/index.d.ts]: I don't think that `Buffer|` is needed at all as `ArrayBufferView` includes `Buffer` - if not the definition of quite some APIs in crypto would be wrong.

**Comment by kjin** [comment]
@ZaneHannanAU @Flarna [`Buffer`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/80d50f06cf6f6e075774c13f5fec8012a31529cb/types/node/index.d.ts#L265) doesn't appear to have an API similar to `ArrayBufferView` at all... unless I'm mistaken? In `lib.es5.d.ts`:

```ts
interface ArrayBufferView {
    /**
      * The ArrayBuffer instance referenced by the array.
      */
    buffer: ArrayBufferLike;

    /**
      * The length in bytes of the array.
      */
    byteLength: number;

    /**
      * The offset in bytes of the array.
      */
    byteOffset: number;
}
```

**Comment by Flarna** [comment]
I'm a little bit confused here. I expected `ArrayBufferView` is the combination of `TypedArray`s, `Buffer` and `DataView` but actually this is not true.
But typescript allows to pass a `Buffer` to an interface expecting an `ArrayBufferView`.
Maybe we should replace `ArrayBufferView` by something like this which would map to API doc:
```
type TypedArray = Int8Array | Uint8Array | Uint8ClampedArray | Int16Array | Uint16Array | Int32Array | Uint32Array | Float32Array | Float64Array;
and then use TypedArray | DataView | Buffer
```

**Comment by ZaneHannanAU** [comment]
@kjin quite completely mistaken.

ArrayBufferView is just an interface. Anything implementing those three (buffer, byteOffset, byteLength) as those types (ArrayBuffer/ArrayBufferLike, number, number) is taken as such.

… basically yeah ArrayBufferView is fine with anything implementing those three specifically. Buffer, Uint8Array, Uint32Array, DataView, BigInt64Array, Float64Array etc.

**Comment by Flarna** [comment]
In crypto APIs no duck typing is used. If you pass an object fitting the `ArrayBufferView` interface by setting the 3 properties makes typescript happy but node throws with something like this:
```
TypeError [ERR_INVALID_ARG_TYPE]: The "buf" argument must be of type ArrayBufferView. Received type object
    at Object.randomFillSync (internal/crypto/random.js:71:11)
```

**Comment by ZaneHannanAU** [comment]
... realistically if it passes the `ArrayBuffer.isView(obj)` then yeah. Hrmm.

**Comment by ZaneHannanAU** [comment]
I'm guessing it's because they must be readonly?

**Comment by Flarna** [comment]
I think we should do the `ArrayBufferView` => `Buffer | TypedArray | DataView` change in a separate PR as it is not directly related to this one.

**Pull request merged**